crates/solidity/inputs/language/bindings/rules.msgb

```diff
@@ -6,178 +8,326 @@ attribute push_symbol = symbol       => type = "push_symbol", symbol = symbol
 
 @contract [ContractDefinition] {
   node @contract.lexical_scope
-  node @contract.defs
+  node @contract.def
+  node @contract.members
+  node @contract.type_members
+
+  edge @contract.lexical_scope -> @contract.members
+  edge @contract.lexical_scope -> @contract.type_members
 }
 
-@contract [ContractDefinition ... @name name: [Identifier] ...] {
-  node def
-  attr (def) node_definition = @name
+@interface [InterfaceDefinition] {
+  node @interface.lexical_scope
+  node @interface.def
+  node @interface.members
+  node @interface.type_members
+
+  edge @interface.lexical_scope -> @interface.members
+  edge @interface.lexical_scope -> @interface.type_members
+}
+
+@library [LibraryDefinition] {
+  node @library.lexical_scope
+  node @library.def
+  node @library.members
+
+  edge @library.lexical_scope -> @library.members
+}
+
+@struct [StructDefinition] {
+  node @struct.lexical_scope
+  node @struct.def
+  node @struct.members
+}
+
+@enum [EnumDefinition] {
+  node @enum.lexical_scope
+  node @enum.def
+  node @enum.members
+}
+
+@function [FunctionDefinition] {
+  node @function.lexical_scope
+  node @function.def
+}
 
-  edge @contract.lexical_scope -> def
-  edge @contract.defs -> def
+@constant [ConstantDefinition] {
+  node @constant.lexical_scope
+  node @constant.def
 }
 
-;; Connect the contract to its containing source unit
+@error [ErrorDefinition] {
+  node @error.lexical_scope
+  node @error.def
+}
+
+@value_type [UserDefinedValueTypeDefinition] {
+  node @value_type.lexical_scope
+  node @value_type.def
+}
+
+@event [EventDefinition] {
+  node @event.lexical_scope
+  node @event.def
+}
+
+;; Top-level definitions...
 @source_unit [SourceUnit ... [SourceUnitMembers
     ...
-    [SourceUnitMember @contract [ContractDefinition]]
+    [SourceUnitMember @unit_member (
+          [ContractDefinition]
+        | [InterfaceDefinition]
+        | [LibraryDefinition]
+        | [StructDefinition]
+        | [EnumDefinition]
+        | [FunctionDefinition]
+        | [ConstantDefinition]
+        | [ErrorDefinition]
+        | [UserDefinedValueTypeDefinition]
+        | [EventDefinition]
+    )]
     ...
 ] ...] {
-  edge @source_unit.defs -> @contract.defs
-  edge @contract.lexical_scope -> @source_unit.lexical_scope
+  edge @unit_member.lexical_scope -> @source_unit.lexical_scope
+
+  ;; ... are available in the file's lexical scope
+  edge @source_unit.lexical_scope -> @unit_member.def
+
+  ;; ... and are exported in the file
+  edge @source_unit.defs -> @unit_member.def
 }
 
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;;; Interface definitions
+;;; Named definitions (contracts, functions, libraries, etc.)
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-@interface [InterfaceDefinition] {
-  node @interface.lexical_scope
-  node @interface.defs
+@contract [ContractDefinition ... @name name: [Identifier] ...] {
+  node def
```

Comment:
Question: what is the difference between the `def` node defined on line 33 (`node @contract.def`) and the other `def` node defined on line 128 (`node def`)? or is it the same thing?

Reply:
They are different nodes in the graph, although they are roughly equivalent. You could say that `@contract.def` is the "external" definition of the contract entity which will connect to the rest of the graph (see the stanza right above this one), while `def` is the internal definition and is associated to the name CST node of the contract. This pattern is used pervasively for all defined entities.

You need both of them: the node related to the `@contract` you use to refer to the contract, from the parent (ie. the source unit). This is the definiens of the identifier. And you also need the `@name` identifier node because that's how you would usually navigate to the contract using a CST cursor.

Reply:
`identifier` could be misleading, since we will have identifiers that act as references (the nodes called `ref` throughout the rules file). I don't mind `definition` though. Most of the node names I adopted from tree-sitter-stack-graphs but I'm all in for better names, always.

One thing to keep in mind is `definiens` will be an attribute of the definition `identifier`. And this will point to the CST contract node, not the graph node. The `@contract.def` graph node keeps no relation to the CST nodes beyond the graph execution. For example:
```
@contract [ContractDefinition ... @name name: [Identifier] ...] {
  node def
  attr (def) node_definition = @name
  attr (def) definiens_node = @contract
  ;; ...
}
```

Reply:
I see. Let's talk about naming in the next stand up then 👍